### ISSUE TITLE
Operator precedence not followed

### ISSUE THREAD

**Issue by peeyush31** [cherow/cherow#282]
I am using v1.6.8 to parse
**Code**

```js
y && 'test' in ar
```

**Actual behavior:**
```js
{
  "type": "Program",
  "sourceType": "script",
  "body": [
    {
      "type": "ExpressionStatement",
      "expression": {
        "type": "BinaryExpression",
        "left": {
          "type": "LogicalExpression",
          "left": {
            "type": "Identifier",
            "name": "y"
          },
          "right": {
            "type": "Literal",
            "value": "test"
          },
          "operator": "&&"
        },
        "right": {
          "type": "Identifier",
          "name": "ar"
        },
        "operator": "in"
      }
    }
  ]
}
```

**Expected behavior:**

The `in` operator should have higher precedence than `&&`

**Comment by KFlash** [comment]
@peeyush31 Isn't this the same as in other parsers if you compare it in ASTExplorer?

**Comment by peeyush31** [comment]
No there is a difference

for esprima
```js
{
  "type": "Program",
  "body": [
    {
      "type": "ExpressionStatement",
      "expression": {
        "type": "LogicalExpression",
        "operator": "&&",
        "left": {
          "type": "Identifier",
          "name": "y",
          "range": [
            0,
            1
          ]
        },
        "right": {
          "type": "BinaryExpression",
          "operator": "in",
          "left": {
            "type": "Literal",
            "value": "test",
            "raw": "'test'",
            "range": [
              5,
              11
            ]
          },
          "right": {
            "type": "Identifier",
            "name": "ar",
            "range": [
              15,
              17
            ]
          },
          "range": [
            5,
            17
          ]
        },
        "range": [
          0,
          17
        ]
      },
      "range": [
        0,
        17
      ]
    }
  ],
  "sourceType": "module",
  "range": [
    0,
    17
  ]
}
```
The other parsers have a same AST subtree order. I have tested with `acorn`

**Comment by KFlash** [comment]
Esprima isn't ESTree spec compat so their AST differ from the original specs. And they also add different properties on various nodes that doesn't exist in the specs - e.g. `each` in for - in iteration. So I wouldn't have looked at their AST.

**Comment by peeyush31** [comment]
the precedence from `cherow` says that first do `&&` and then operate `in`. This is not correct according to ES standard. the result expression can be rewritten as
```js
(y && 'test') in ar
```
according to cherow. This is the issue

**Comment by KFlash** [comment]
cc/ @fkleuver Can you investigate the precedence order?

**Comment by fkleuver** [comment]
```ts
    const delta = ((t === Token.Exponentiate) as any) << Token.PrecStart;
    if (prec + delta <= minPrec) break;
```

I've always found those 2 lines a bit strange. I'm not sure why you are adding the precedence value of the exponentiation operator to the delta.

Anyway, the `InKeyword` token has a higher precedence value than the `LogicalAnd` token so they've got the correct values. Either the associativity is wrong in the expression parser logic or something else is going on, i'll take a look

**Comment by peeyush31** [comment]
oh sorry, that was the actual output. and the expected will be what you just mentioned

**Comment by fkleuver** [comment]
@KFlash I can confirm cherow parses in the wrong precedence order in his example

**Comment by fkleuver** [comment]
Talk about commenting at the exact same second.. :)

**Comment by KFlash** [comment]
@fkleuver Ok. Is it enough you think to fix the precedence value in token.ts or what you suggest`?

**Comment by fkleuver** [comment]
@KFlash Actually this is a 1.6 bug, you already seem to have fixed it in 2.0.

In 1.6 there's this code:
```ts
    const bit = context & Context.AllowIn ^ Context.AllowIn;
    while (hasBit(parser.token, Token.IsBinaryOp)) {
        const t: Token = parser.token;
        const prec = t & Token.Precedence;
        const delta = ((t === Token.Exponentiate) as any) << Token.PrecStart;
        if (bit && t === Token.InKeyword) break;
        if (prec + delta <= minPrec) break;
        nextToken(parser, context);
```

It does not "allow in" at that point, so it breaks out and that messes up the precedence. But this code is not present in 2.0. I'm not sure however if removing that particular line in 1.6 won't break other stuff. Did you write that code or was it someone else?

**Comment by peeyush31** [comment]
just for help
the test case with 
```js
y * x + z
```
is working correctly.

**Comment by KFlash** [comment]
@fkleuver 
Here is the basic ideas behind the precedence code. I can't look into that now, but can you take a look?
Main idea was to make the "precedence" conditionally higher without branching by having `Context.AllowIn` a higher bit than token precendence bits and doing `state.token & Token.Precedence & bit` instead, where `const bit = context & Context.AllowIn ^ Context.AllowIn`.
Then just compare `prec + (this.token === Token.Exponentiate) > minPrec `- I verified it does compile to a cmpl + setzl, so it's fully branch-free.
After that was to add an extra bit to `Token.LogicalAnd and Token.LogicalOr to tell them apart, so we're only doing a bit test instead of two equality checks.

**Comment by KFlash** [comment]
I'm aware of that and what you point out is causing a bug as we know. But fixing that in 1.6.x now will break too much I guess. So in 2.0 I'm allowing AllowIn by default and disallow it in a few places. 
Howver. This `if ((prec | bit) + delta <= minPrec) break;` never worked as planned. It seems to be related to `token.ts`. Can you investigate that? I added precedence tests now so should be easy to track

**Comment by fkleuver** [comment]
> But fixing that in 1.6.x now will break too much I guess.

Break what? Tests? Can't be that many..

Surely can't breaks apps.. (that's what it's doing right now because of the bug lol)

> if ((prec | bit) + delta <= minPrec) break;

How do you intend it to work? It's supposed to accomplish right-to-left associativity for the exponentiation operator right?

**Comment by KFlash** [comment]
Yes. Right to left. It should break the loop if the AllowIn bit get higher / lower than `Token.Precedence`. 
As I wrote earlier. The "precedence"  has to be conditionally higher without branching by having Context.AllowIn a higher bit than token precendence bits. As it is now it seems like either the precedence tokens or the context masks are out of order.

Well. I fix it in 1.6x soon after I'm done with 2.0

**Comment by fkleuver** [comment]
Regarding this

```ts
const bit = context & Context.AllowIn ^ Context.AllowIn;
const prec = t & Token.Precedence;
const delta = (t === Token.Exponentiate) << Token.PrecStart;
if ((prec | bit) + delta <= minPrec) break;
// ...
right: this.parseBinaryExpression(context, prec, this.startNode()),
```

The precedence for the tokens is fine, they correctly go in ascending order from `LogicalOr` to `Exponentiate`. The context is also fine. `AllowIn` is 5 bits higher than the highest possible precedence.

This is where the error seems to be:
```ts
if ((prec | bit) + delta <= minPrec) break;
```

The `bit` is on the left hand side of a less-than, meaning it accomplishes the exact opposite of what it needs to.

This would probably work:
```ts
if (prec + delta - bit <= minPrec) break;
```

Then if `bit` is set you get a negative value and it's guaranteed to break.

**Comment by fkleuver** [comment]
You remembered to remove the `^` right? :P

**Comment by KFlash** [comment]
[this](https://github.com/cherow/cherow/blob/beta/src/state.ts#L1433) is the 2.0 solution and tests pass locale

**Comment by fkleuver** [comment]
Why did you break the tests :(

**Comment by fkleuver** [comment]
What do you think about this :) https://github.com/cherow/cherow/pull/275/commits/a8fc459e4eb2967afbbc777b670011e7c2591100

**Comment by KFlash** [comment]
@fkleuver The tests pass so it's a good solution:)  And less code and branching! Well done 👍

**Comment by KFlash** [comment]
@fkleuver I guess this is solved in v. 2.0 and the beta will be out soon.  Do we still need to fix this in 1.6x ?

**Comment by KFlash** [comment]
Fix is out on current Master branch. So closing.
